pkg/util/machine/azureproviderspec_test.go

```diff
@@ -0,0 +1,152 @@
+	},
+	Location:             "neversayneverland",
+	NetworkResourceGroup: "networkRG",
+	OSDisk: azureproviderv1beta1.OSDisk{
+		DiskSizeGB: int32(9001),
+		OSType:     "Linux",
+		ManagedDisk: azureproviderv1beta1.ManagedDiskParameters{
+			StorageAccountType: "Premium_LRS",
+		},
+	},
+	PublicIP:           false,
+	PublicLoadBalancer: "",
+	ResourceGroup:      "myRG",
+	VMSize:             "Standard_D8s_v3",
+	Vnet:               "myVnet",
+	Zone:               to.StringPtr("2"),
+}
+
+func marshalAzureMachineProviderSpec(t *testing.T, spec kruntime.Object) []byte {
+	serializer := kjson.NewSerializerWithOptions(
+		kjson.DefaultMetaFactory, scheme.Scheme, scheme.Scheme,
+		kjson.SerializerOptions{Yaml: false},
+	)
+
+	json := scheme.Codecs.CodecForVersions(serializer, nil, schema.GroupVersions(scheme.Scheme.PrioritizedVersionsAllGroups()), nil)
+
+	buf := &bytes.Buffer{}
+	err := json.Encode(spec, buf)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return buf.Bytes()
+}
+
+func TestUnmarshalAzureProviderSpec(t *testing.T) {
+	machineName := "myMachine"
+
+	// Register old type soley so we can encode the CR to bytes for use only in test
+	scheme.Scheme.AddKnownTypes(azureproviderv1beta1.SchemeGroupVersion, &azureproviderv1beta1.AzureMachineProviderSpec{})
+
+	machineProviderSpecBytes := marshalAzureMachineProviderSpec(t, machineAzureProviderSpec)
+	azProviderSpecBytes := marshalAzureMachineProviderSpec(t, azAzureProviderSpec)
+
+	for _, tt := range []struct {
+		name         string
+		machineType  MachineType
+		providerSpec []byte
+		wantErr      string
+	}{
+		{
+			name:         "valid - machine.openshift.io provider spec",
+			providerSpec: machineProviderSpecBytes,
+		},
+		{
+			name:         "valid - azureproviderconfig.openshift.io provider spec",
+			providerSpec: azProviderSpecBytes,
+		},
+		{
+			name:         "fail - azureproviderconfig.openshift.io invalid json",
+			providerSpec: []byte("azureproviderconfig.openshift.io"),
+			machineType:  MachineSet,
+			wantErr:      fmt.Sprintf("%s %s: failed to unmarshal the 'azureproviderconfig.openshift.io' provider spec: %q", MachineSet, machineName, "invalid character 'a' looking for beginning of value"),
+		},
+		{
+			name:         "fail - unable to decode type",
+			providerSpec: []byte(`{"apiVersion": "invalid/v1beta1","kind": "AzureMachineProviderSpec"}`),
+			machineType:  MachineSet,
+			wantErr:      `no kind "AzureMachineProviderSpec" is registered for version "invalid/v1beta1" in scheme`,
+		},
+		{
+			name:         "fail - invalid type",
+			providerSpec: []byte(`{"apiVersion": "v1","kind": "Node"}`),
+			machineType:  MachineSet,
+			wantErr:      `failed to read provider spec:`,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			actualProviderSpec, err := UnmarshalAzureProviderSpec(machineName, tt.machineType, tt.providerSpec)
+			if err == nil {
+				if tt.wantErr != "" {
+					t.Error(err)
+				}
+
+				// Don't care about type meta because it should retain it
```

Comment:
Minor. Maybe this should be something like:

// Ignoring check for support for both azureproviderconfig and machine providerSpec Types